File: design-principles/law-of-demeter/go/LawOfDemeter.go
/**
 * 迪米特法则（Law of Demeter，LoD）
 * 也叫最少知识原则（Principle of Least Knowledge）。
 * 即一个对象应当对其他对象尽可能少的了解，对象之间的耦合度应该尽可能低。
 */
package main

import (
  "fmt"
)

// Product类，表示商品
type Product struct {
  name  string
  price float64
}

func NewProduct(name string, price float64) *Product {
  return &Product{name: name, price: price}
}

func (p *Product) GetName() string {
  return p.name // 获取商品名称
}

func (p *Product) GetPrice() float64 {
  return p.price // 获取商品价格
}

// ShoppingCart类，表示购物车
type ShoppingCart struct {
  products []*Product // 存储商品的切片
}

func NewShoppingCart() *ShoppingCart {
  return &ShoppingCart{products: []*Product{}}
}

func (sc *ShoppingCart) AddProduct(product *Product) {
  sc.products = append(sc.products, product) // 添加商品到购物车
}

func (sc *ShoppingCart) CalculateTotalPrice() float64 {
  totalPrice := 0.0
  for _, product := range sc.products {
    totalPrice += product.GetPrice() // 计算购物车中所有商品的总价格
  }
  return totalPrice
}

func (sc *ShoppingCart) ValidateProduct(product *Product) bool {
  return product.GetPrice() < 1000
}

// Customer类，表示顾客
type Customer struct {
  name         string        // 顾客名称
  shoppingCart *ShoppingCart // 顾客的购物车
}

func NewCustomer(name string) *Customer {
  return &Customer{name: name, shoppingCart: NewShoppingCart()}
}

// 购买商品方法
func (c *Customer) Buy(product *Product) {
  // 只跟购物车打交道，且不用了解购物车实现细节，只是调用购物车方法
  if c.shoppingCart.ValidateProduct(product) {
    c.shoppingCart.AddProduct(product)                 // 将商品添加到购物车
    totalPrice := c.shoppingCart.CalculateTotalPrice() // 计算总价格
    fmt.Printf("%s 购买了 %s 价格为 %.2f\n", c.name, product.GetName(), totalPrice)
  } else {
    fmt.Printf("%s 的价格超出范围：%.2f\n", product.GetName(), product.GetPrice())
  }
}

func main() {
  customer := NewCustomer("Jimmy")
  product1 := NewProduct("Computer", 5000)
  product2 := NewProduct("Book", 200)

  customer.Buy(product1)
  customer.Buy(product2)
}

/**
jarry@jarrys-MacBook-Pro law-of-demeter % go run go/LawOfDemeter.go
Computer 的价格超出范围：5000.00
Jimmy 购买了 Book 价格为 200.00
*/

File: design-principles/law-of-demeter/python/LawOfDemeter.py
"""
 * 迪米特法则（Law of Demeter，LoD）
 * 也叫最少知识原则（Principle of Least Knowledge）。
 * 即一个对象应当对其他对象尽可能少的了解，对象之间的耦合度应该尽可能低。
"""
# Product类，表示商品


class Product:
    def __init__(self, name, price):
        self.name = name  # 商品名称
        self.price = price  # 商品价格

    def get_name(self):
        return self.name  # 获取商品名称

    def get_price(self):
        return self.price  # 获取商品价格


# ShoppingCart类，表示购物车
class ShoppingCart:
    def __init__(self):
        self.products = []  # 存储商品的列表

    def add_product(self, product):
        self.products.append(product)  # 添加商品到购物车

    def calculate_total_price(self):
        # 计算购物车中所有商品的总价格
        return sum(product.get_price() for product in self.products)

    def validate_product(self, product):
        return product.get_price() < 1000


# Customer类，表示顾客
class Customer:
    def __init__(self, name):
        self.name = name  # 顾客名称
        self.shopping_cart = ShoppingCart()  # 顾客的购物车

    # 只跟购物车打交道，且不用了解购物车实现细节，只是调用购物车方法
    def buy(self, product):
        # 购买商品方法
        if self.shopping_cart.validate_product(product):
            self.shopping_cart.add_product(product)  # 将商品添加到购物车
            total_price = self.shopping_cart.calculate_total_price()  # 计算总价格
            print(f"{self.name} 购买了 {product.get_name()} 价格为 {total_price}")
        else:
            print(f"{product.get_name()} 的价格超出范围：{product.get_price()}")


"""
 * 测试符合迪米特原则的例子
 * 对象职责单一清晰，不把很多功能堆积在一起。
 * 拆出一个购物车类，作为顾客和商品之间的中介类。
 * 对象不直接依赖朋友的朋友。
"""
customer = Customer("Jimmy")
customer.buy(Product("Computer", 5000))
customer.buy(Product("Book", 200))

"""
jarry@jarrys-MacBook-Pro law-of-demeter % python python/LawOfDemeter.py 
Computer 的价格超出范围：5000
Jimmy 购买了 Book 价格为 200
"""